go/vt/vtgate/planbuilder/operators/derived.go

```diff
@@ -132,28 +134,41 @@ func (d *Derived) AddPredicate(ctx *plancontext.PlanningContext, expr sqlparser.
 	return d, nil
 }
 
-func (d *Derived) AddColumn(ctx *plancontext.PlanningContext, expr sqlparser.Expr) (int, error) {
-	col, ok := expr.(*sqlparser.ColName)
+func (d *Derived) AddColumn(ctx *plancontext.PlanningContext, expr *sqlparser.AliasedExpr, reuseCol bool) (ops.Operator, int, error) {
+	col, ok := expr.Expr.(*sqlparser.ColName)
 	if !ok {
-		return 0, vterrors.VT13001("cannot push non-colname expression to a derived table")
+		return nil, 0, vterrors.VT13001("cannot push non-colname expression to a derived table")
+	}
+
+	if reuseCol {
+		for offset, column := range d.Columns {
+			if ctx.SemTable.EqualsExpr(col, column) {
+				return d, offset, nil
+			}
+		}
 	}
```

Comment:
We're using the same small logic in `ApplyJoin.AddColumn`, how do you feel about extracting this to a common function?

```suggestion
	if offset, ok := myExtractedFindColFunc(ctx, reuseCol, d.Columns, col); ok {
		return d, offset, nil
	}
```

```go
func myExtractedFindColFunc(ctx, reuseCol, d.Columns, col) (int, bool) {
	if !reuseCol {
		return 0, false
	}

	for offset, column := range d.Columns {
		if ctx.SemTable.EqualsExpr(col, column) {
			return offset, true
		}
	}
	return 0, false
}
```